port/cpl_float.h

```diff
@@ -50,13 +50,576 @@
+            {
+                // Plus or minus zero.
+                return unsigned2float(iSign << 31);
+            }
+
+            // Denormalized number -- renormalize it.
+            while (!(iMantissa & 0x00000400))
+            {
+                iMantissa <<= 1;
+                iExponent -= 1;
+            }
+
+            iExponent += 1;
+            iMantissa &= ~0x00000400U;
+        }
+
+        // Normalized number.
+        iExponent = iExponent + (127 - 15);
+        iMantissa = iMantissa << 13;
+
+        // Assemble sign, exponent and mantissa.
+        /* coverity[overflow_sink] */
+        return unsigned2float((iSign << 31) |
+                              (static_cast<std::uint32_t>(iExponent) << 23) |
+                              iMantissa);
+    }
+
+    template <typename T> static repr toRepr(T fValue)
+    {
+        return computeToRepr(static_cast<compute>(fValue));
+    }
+
+    template <typename T> static T fromRepr(repr rValue)
+    {
+        return static_cast<T>(reprToCompute(rValue));
+    }
+
+#endif
+
+  private:
+    repr rValue;
+
+  public:
+    compute get() const
+    {
+        return reprToCompute(rValue);
+    }
+
+    GFloat16() = default;
+    GFloat16(const GFloat16 &) = default;
+    GFloat16(GFloat16 &&) = default;
+    GFloat16 &operator=(const GFloat16 &) = default;
+    GFloat16 &operator=(GFloat16 &&) = default;
+
+    // Constructors and conversion operators
+
+#ifdef HAVE__FLOAT16
+    // cppcheck-suppress noExplicitConstructor
+    constexpr GFloat16(_Float16 hfValue) : rValue(hfValue)
+    {
+    }
+
+    constexpr operator _Float16() const
+    {
+        return rValue;
+    }
+#endif
+
+    // cppcheck-suppress-macro noExplicitConstructor
+#define GDAL_DEFINE_CONVERSION(TYPE)                                           \
+                                                                               \
+    GFloat16(TYPE fValue) : rValue(toRepr(fValue))                             \
+    {                                                                          \
+    }                                                                          \
+                                                                               \
+    operator TYPE() const                                                      \
+    {                                                                          \
+        return fromRepr<TYPE>(rValue);                                         \
+    }
+
+    GDAL_DEFINE_CONVERSION(float)
+    GDAL_DEFINE_CONVERSION(double)
+    GDAL_DEFINE_CONVERSION(char)
+    GDAL_DEFINE_CONVERSION(signed char)
+    GDAL_DEFINE_CONVERSION(short)
+    GDAL_DEFINE_CONVERSION(int)
+    GDAL_DEFINE_CONVERSION(long)
+    GDAL_DEFINE_CONVERSION(long long)
+    GDAL_DEFINE_CONVERSION(unsigned char)
+    GDAL_DEFINE_CONVERSION(unsigned short)
+    GDAL_DEFINE_CONVERSION(unsigned int)
+    GDAL_DEFINE_CONVERSION(unsigned long)
+    GDAL_DEFINE_CONVERSION(unsigned long long)
```

Comment:
we should definitely not just use plain "char" as depending on the platform it might be signed or unsigned. More generally, why not using stdint types [u]int[8/16/32/64] ?

Reply:
That's a C++ thing... I want to allow conversions from/to all other types, similar to how `float` behaves. Even though `char` is either signed or unsigned, it's a different type from both `signed char` and `unsigned char`. Similarly, the `std::[u]int*` types are defined to be aliases of other types, but it's not defined which other types these are.

For example, it can that e.g. `std::uint8_t` is defined to be `char` (assuming that is an unsigned type), and when then someone converts from/to `unsigned char`, there would be an error. Similarly, `std::int64_t` is usually either `long` or `long long`, but it's not clear which (assuming that `long` has 64 bits). If `std::int64_t` is `long long, then if someone converts from `1L` there would be an error (and vice versa with `long` and `1LL`). Even though `long` and `long long` maybe behave identically, they are considered different types.

For example, gdal uses both `std::int64_t` and `long long` (via `GInt64`) internally. These might be different types. (I think they are the same on Linux but different on macOS.)

It's good practice in C++ to define overloads only for the basic types, not for the `std::[u]int*` aliases.

Reply:
ok, that makes sense (and I initially missed you had actually a instance for ``unsigned char``)